File: app/src/main/java/com/codepath/shapes/shape/Shape.java
package com.codepath.shapes.shape;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;

/**
 * Created by JaneChung on 3/7/17.
 */

public abstract class Shape {
    protected Paint mPaint;
    private float mXPosition;
    private float mYPosition;
    @ColorInt
    private int mColor;

    public Shape(float x, float y, @ColorInt int color) {
        mXPosition = x;
        mYPosition = y;
        mColor = color;
        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mPaint.setColor(mColor);

    }

    public float getX() {
        return mXPosition;
    }

    public void setX(float x) {
        mXPosition = x;
    }

    public float getY() {
        return mYPosition;
    }

    public void setY(float y) {
        mYPosition = y;
    }

    @ColorInt
    public int getColor() {
        return mColor;
    }

    public void setColor(@ColorInt int color) {
        mColor = color;
        mPaint.setColor(mColor);
    }

    public abstract void draw(@NonNull Canvas canvas);

    //serialize method ?

}

File: app/src/main/java/com/codepath/shapes/shape/Shape.kt
package com.codepath.shapes.shape

import android.graphics.Canvas
import android.graphics.Color.BLACK
import android.graphics.Paint
import android.graphics.RectF
import android.support.annotation.ColorInt
import com.codepath.shapes.util.setCenterAndDimensions

sealed class Shape(var x: Float = 0f, var y: Float = 0f, @ColorInt color: Int = BLACK) {
    protected var mPaint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)

    init {
        mPaint.color = color
    }

    var color: Int
        get() = mPaint.color
        set(@ColorInt color) {
            mPaint.color = color
        }

    abstract fun draw(canvas: Canvas)
}

class Rectangle(x: Float, y: Float, @ColorInt color: Int, var width: Float, var height: Float) : Shape(x, y, color) {
    // Recycled RectF for drawing
    private val mRectF = RectF()

    override fun draw(canvas: Canvas) {
        mRectF.setCenterAndDimensions(x, y, width, height)
        canvas.drawRect(mRectF, mPaint)
    }
}

class Circle(x: Float, y: Float, @ColorInt color: Int, var radius: Float) : Shape(x, y, color) {
    override fun draw(canvas: Canvas) = canvas.drawCircle(x, y, radius, mPaint)
}

class Text(x: Float, y: Float, @ColorInt color: Int, textSize: Float, var text: String) : Shape(x, y, color) {
    init {
        mPaint.textSize = textSize
    }

    var fontSize: Float
        get() = mPaint.textSize
        set(size) {
            mPaint.textSize = size
        }

    override fun draw(canvas: Canvas) = canvas.drawText(text, 0, text.length, x, y, mPaint)
}
